protocol/x/accountplus/ante/ante.go

```diff
@@ -0,0 +1,299 @@
+	"github.com/dydxprotocol/v4-chain/protocol/x/accountplus/types"
+)
+
+// AuthenticatorDecorator is responsible for processing authentication logic
+// before transaction execution.
+type AuthenticatorDecorator struct {
+	accountPlusKeeper *keeper.Keeper
+	accountKeeper     authante.AccountKeeper
+	sigModeHandler    *txsigning.HandlerMap
+	cdc               codec.Codec
+}
+
+// NewAuthenticatorDecorator creates a new instance of AuthenticatorDecorator with the provided parameters.
+func NewAuthenticatorDecorator(
+	cdc codec.Codec,
+	accountPlusKeeper *keeper.Keeper,
+	accountKeeper authante.AccountKeeper,
+	sigModeHandler *txsigning.HandlerMap,
+) AuthenticatorDecorator {
+	return AuthenticatorDecorator{
+		accountPlusKeeper: accountPlusKeeper,
+		accountKeeper:     accountKeeper,
+		sigModeHandler:    sigModeHandler,
+		cdc:               cdc,
+	}
+}
+
+// AnteHandle is the authenticator ante handler responsible for processing authentication
+// logic before transaction execution.
+func (ad AuthenticatorDecorator) AnteHandle(
+	ctx sdk.Context,
+	tx sdk.Tx,
+	simulate bool,
+	next sdk.AnteHandler,
+) (newCtx sdk.Context, err error) {
+	defer metrics.ModuleMeasureSince(
+		types.ModuleName,
+		metrics.AuthenticatorDecoratorAnteHandleLatency,
+		time.Now(),
+	)
+
+	// Make sure smart account is active.
+	if active := ad.accountPlusKeeper.GetIsSmartAccountActive(ctx); !active {
+		return ctx, types.ErrSmartAccountNotActive
+	}
+
+	// Authenticators don't support manually setting the fee payer
+	err = ad.ValidateAuthenticatorFeePayer(tx)
+	if err != nil {
+		return sdk.Context{}, err
+	}
+
+	msgs := tx.GetMsgs()
+	if len(msgs) == 0 {
+		return ctx, errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "no messages in transaction")
+	}
+
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, errorsmod.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	// The fee payer is the first signer of the transaction. This should have been enforced by the
+	// LimitFeePayerDecorator
+	signers, _, err := ad.cdc.GetMsgV1Signers(msgs[0])
+	if err != nil {
+		return ctx, errorsmod.Wrap(sdkerrors.ErrUnauthorized, "failed to get signers")
+	}
+	feePayer := sdk.AccAddress(signers[0])
+	feeGranter := feeTx.FeeGranter()
+	fee := feeTx.GetFee()
+
+	selectedAuthenticators, err := ad.GetSelectedAuthenticators(tx, len(msgs))
+	if err != nil {
+		return ctx, err
+	}
+
+	// tracks are used to make sure that we only write to the store after every message is successful
+	var tracks []func() error
+
+	// Authenticate the accounts of all messages
+	for msgIndex, msg := range msgs {
+		signers, _, err := ad.cdc.GetMsgV1Signers(msg)
+		if err != nil {
+			return ctx, errorsmod.Wrap(sdkerrors.ErrUnauthorized, "failed to get signers")
+		}
+		// Enforce only one signer per message
+		if len(signers) != 1 {
+			return sdk.Context{}, errorsmod.Wrap(sdkerrors.ErrUnauthorized, "messages must have exactly one signer")
+		}
+
+		// By default, the first signer is the account that is used
```

Comment:
Nit: updat to reflect that the above guarantees there is one and one signer only
```suggestion
		// Get the account corresponding to the only signer of this message
```